### DISCUSSION TITLE
Error with CreateProduct - SDK PHP

### DISCUSSION THREAD

**Discussion by svilapp-sviluppo** [fattureincloud/api#30] [Q&A] [answered]
Hello,
I started to convert your API from v1 to v2 and in creating the product I get an incomprehensible error.

Below is the value of the array I send to `CreateProductRequest`:
```
array(8) {
  ["code"]=>
  string(6) "TEST-1"
  ["name"]=>
  string(23) "New Product Test 232"
  ["description"]=>
  string(3) "zxc"
  ["use_gross_price"]=>
  bool(false)
  ["net_price"]=>
  float(0.8)
  ["gross_price"]=>
  float(0.98)
  ["in_stock"]=>
  bool(true)
  ["stock_initial"]=>
  float(110)
}
```
I get an ApiException with this message:
```
[422] Client error: `POST https://api-v2.fattureincloud.it/c/XXXXX/products` resulted in a `422 Unprocessable Entity` response: {"error":{"message": "Invalid request.", "validation_result":{"data":["The i data field is required."]}}}
```

I honestly can't figure out where the error is, I looked at the documentation and API Reference without being able to figure it out so I decided to ask you directly.

Thank you for your help.

**Comment by valmoz** [comment]
Hi @svilapp-sviluppo 
this error is usually returned when a badly formatted request is sent. I know, not the best error message, we're trying to improve it.

It is strange though because you're currently using an SDK, so you should be able to avoid this kind of error. To indagate more in detail I need more info, so I suggest you contact Customer Support on the Fatture in Cloud page. You can send them the link to this conversation, so they'll know the current situation.

Thanks
Mauro

**Comment by svilapp-sviluppo** [comment]
Hello @valmoz ,
in fact I also tried to analyze the various models but without an understandable match. Thanks for the suggestion.
Regards.

**Reply by valmoz** [reply]
Yes, that's why I need to investigate in more detail, to understand if there's a problem with the SDK or with the API. :)
The Customer Support will surely come back to me, but at least I'll obtain more data that we cannot share in this public discussion.

**Reply by valmoz** [reply]
Hi @svilapp-sviluppo 
we checked, and it seems that you're not sending the body of the request. We receive an empty array.
Please check to set the body correctly while executing the request.

**Reply by svilapp-sviluppo** [reply]
Sorry but I use your SDK:
```
$createProductRequest = new CreateProductRequest($data);

try {
    $result = Api::getProducts()->createProduct($companyId, $createProductRequest);
    print_r($result);
} catch (ApiException $e) {
    echo $e->getMessage();
    exit;
}
```

**Reply by valmoz** [reply]
Hi @svilapp-sviluppo 
I just tried this code and it works:
```php
<?php
use FattureInCloud\Api\ProductsApi;
use FattureInCloud\Model\CreateProductRequest;
use FattureInCloud\Model\Product;
use FattureInCloud\Configuration;
use GuzzleHttp\Client;

require __DIR__ . '/../vendor/autoload.php';
session_start();

// Retrieve the access token from the session variable
$accessToken = 'ACCESS_TOKEN'; 

// Get the API config and construct the service object.
$config = Configuration::getDefaultConfiguration()->setAccessToken($accessToken);

$productsApi = new ProductsApi(
    new Client(),
    $config
);

try {
    $prod = new Product();
    $prod->setName("bbb");
    $body = new CreateProductRequest();
    $body->setData($prod);
    // Retrieve the first company id
    $product = $productsApi->createProduct("COMPANY_ID", $body);
    print($product);

} catch (Exception $e) {
    echo 'Exception when calling the API: ', $e->getMessage(), PHP_EOL;
}

```

**Reply by svilapp-sviluppo** [reply]
I found the mistake inside `CreateProductRequest` class , the `__construct` accept only Array but setData accept `Product|null` model for that i pass the product data array inside the `_construct` instead the `Product` object. I think you need to make sync about this 2 method.
in the end:
- construct accept array
- setData accept product|null

**Comment by valmoz** [comment, marked as answer]
Summary of the discussion:
that specific error is returned when the JSON has a different structure than expected, checking the request we found out that an empty array was passed.

The problem was related to the way the body request was initialized. the example code passed in the description instead works correctly.
